reactionimp/icemachinesplus
# Fix for "Data injection in the initial product pages and tag pages."

This PR closes issue #117

Changes:


Resolves #issueNumber  
Impact: **breaking|critical|major|minor**  
Type: **feature|bugfix|performance|test|style|refactor|docs|chore**

## Issue
Don't show the loading spinner for PDP. Instead wait until all data is available before we start the hydrate process.

## Solution
- Introduce a two-pass render environment variable that allows to have
the client subscribe for all subscriptions. This render phase is done in
an invisible container (id="react-root-tmp"). After all subscriptions
are in place, we trigger the actual hydration of the server rendered
DOM with all data in place.

## Breaking changes
none


## Testing
1. Set  "REACTION_TWO_PASS_RENDERING": true in settings.json
2. Navigate to a product detail page, e.g. http://localhost:3000/product/manitowoc-qm-45a
3. No loading spinner should be observable.

## Notice
Before the second render pass kicks in, a reordering of the related products can be observed. This is probably not related to React rendering, because the first pass is rendered invisible anyway. I guess that would have to be optimized in the server rendered HTML, which is certainly doable if one throws more effort into it.

## Screencasts

One pass rendering
![one-pass-render](https://user-images.githubusercontent.com/1733229/38676469-6c43ee2e-3e5b-11e8-99e5-8c713736d33e.gif)


Two pass rendering
![two-pass-render](https://user-images.githubusercontent.com/1733229/38676419-3df59658-3e5b-11e8-80e4-b454955fb818.gif)
